Conversion of Class to Functional component

jason-vasilev opened the pull request:
## Summary

This PR started in March 2024 as a straight class → function-component
conversion for the two components in the app (`NewsFeed`, `NewsCard`).
While it sat open, the broader project tooling drifted out of date, so
the scope has since grown to bring the repo to a current, maintained
baseline. The functional-component work remains the foundation; the
newer commits build on top of it.

## 1. Class → function components (original scope)

- `NewsFeed` and `NewsCard` converted from class components to
  function components using `useState` / `useEffect`.
- Removed lifecycle boilerplate and `this`-bound handlers.
- Renamed the loading flag and tightened its typing for clarity.

Goals preserved from the original description:
- **Improve performance** — function components avoid class instance
  overhead and play well with future React optimizations (concurrent
  features, RSC-ready code style).
- **Reduce complexity** — no more `this`, no more `bind`, smaller
  surface area per component.
- **Follow trend** — hooks are the idiomatic React API and classes are
  no longer recommended by the React team.

## 2. Build tooling: CRA → Vite, yarn → npm

- Removed `react-scripts` and the entire CRA tool-chain. Create React
  App was deprecated by the React team in March 2023 and formally
  sunset in February 2025, so it no longer makes sense to ship with it.
- Migrated to **Vite 8** with `@vitejs/plugin-react`. Dev server
  starts near-instantly and the production build uses esbuild +
  Rollup. Config lives in a readable `vite.config.js` (no ejecting).
- Switched the package manager from **yarn** to **npm**. `yarn.lock`
  removed; `package-lock.json` is ignored per project preference.
- Dependencies refreshed: React 18.3, TypeScript 6, Vite 8, Sass 1.99,
  stylelint 17 with `stylelint-config-standard-scss` 17.
- `tsconfig.json` tightened to `strict: true`; production build runs
  `tsc -b && vite build`.

## 3. Hook bug fixes discovered during the conversion

Reviewing the newly-functional components surfaced two real bugs that
pre-existed the conversion and are fixed here:

- **`NewsFeed`: state was being mutated, not set.** The destructures
  were missing their setters, so code like
  `randomStoriesContent.push(data)` and `randomStoriesData = storyUrls`
  never notified React. Rendering only worked by accident because
  `setLoading("")` happened to trigger a final re-render. Fixed by
  using real setters, replacing push/sort mutation with fresh arrays,
  and wrapping the fetch chain in an `AbortController` so the effect
  cleans up on unmount — which also prevents duplicate items under
  React 18 StrictMode's double-invocation in dev.
- **`NewsCard`: author karma fetched on every mouseover.** Hovering
  a card's author link fired a new network request each time. Now
  gated by a `useRef` flag so each card fetches at most once, with an
  abortable fetch on unmount. Also tightened prop types
  (`any` → `NewsCardProps`) and hoisted pure helpers (`timeSince`,
  `imageId`) out of the render body.

## 4. SCSS: `@import` → `@use` / `@forward`

Dart Sass deprecated `@import` and will remove it in Sass 3.0, so the
stylesheets are migrated to the modern module system:

- `src/styles/main.scss` now uses `@forward` as a barrel.
- Partials (`_mixins`, `_typography`, `_grid`) explicitly `@use` their
  dependencies (`vars`, `mixins`) since the module system is
  namespaced.
- Component stylesheets (`NewsCard.scss`, `NewsFeed.scss`) `@use` only
  the partials they need rather than pulling in all of `main.scss` —
  this also eliminates a pre-existing duplication where each component
  chunk was re-emitting all global CSS.

### Other stylesheet cleanup

- Removed duplicate declarations in `_normalize.scss` left over from
  stripped vendor prefixes (`text-size-adjust`, `height: fit-content`,
  `width: fit-content`).
- `::input-placeholder` (invalid) → `::placeholder`.
- `appearance: button` (deprecated) → `appearance: auto`.
- `word-break: break-word` (deprecated) → `overflow-wrap: break-word`.
- Renamed `stripUnit` → `strip-unit` to satisfy the SCSS kebab-case
  function-name rule.

## 5. Docs

- `README.md` rewritten to describe the current stack, the Vite/npm
  migration, available scripts, project structure, styling
  conventions, and the Hacker News API endpoints used.

## Test plan

- [ ] `npm install` from a clean `node_modules`
- [ ] `npm run dev` — app loads at `http://localhost:3000`, 10 random
      stories render once, sorted by score (no duplicates under
      StrictMode)
- [ ] Hover an author link — karma appears in the link's `title`
      attribute; network panel shows exactly one `/user/...` request
      per card no matter how many times you hover
- [ ] `npm run build` — completes without TypeScript errors and
      without Sass or Lightning CSS warnings
- [ ] `npm run preview` — the production build renders correctly
- [ ] `npm run lint:style` — exits 0

jason-vasilev committed:
Half-way conversion class - functional component

jason-vasilev committed:
Clean up

jason-vasilev committed:
Success, clean up and comment

alordiel commented on src/index.tsx:
```diff
@@ -4,13 +4,13 @@ import NewsFeed from './NewsFeed/NewsFeed';
 import reportWebVitals from './reportWebVitals';
 
 ReactDOM.render(
```
This is React 17, may be good thing to migrate to 18? there you have another way to render ([old one](https://react.dev/reference/react-dom/render)  vs [new one](https://react.dev/reference/react-dom/client/createRoot)), it is a bit cleaner.

alordiel commented on src/index.tsx:
```diff
@@ -4,13 +4,13 @@ import NewsFeed from './NewsFeed/NewsFeed';
-		</main>
-	</React.StrictMode>,
-	document.getElementById('root')
+  <React.StrictMode>
```
For cleaner code you can assign this into an <App> component that you can pass later to the render function

alordiel commented on src/NewsFeed/NewsFeed.tsx:
```diff
@@ -5,11 +5,15 @@ import NewsCard from "../NewsCard/NewsCard";
+   * 1 - Almost there
+   * 2 - show content
+   */
+  let [isLoading, setLoading] = useState(2);
```
This is more an issue with the naming. When using `isSomething` I would expect that the variable will have boolean value - true or false. In this case it has 3 values, so I would call it `loadingStatus` and instead of having the numeric values I will use the 'loading', 'almost-there', 'done'. This comes from Clean Code concepts where it is recommended to have more of a verbal values on variables like this then using numeric ones and have a comment somewhere that explains what is what.

alordiel replied:
And you are missing the type of that state.

jason-vasilev replied:
Thanks. That was valuable. I'll have a deep look at Clean Code concepts.

alordiel commented on src/NewsFeed/NewsFeed.tsx:
```diff
@@ -5,11 +5,15 @@ import NewsCard from "../NewsCard/NewsCard";
+   */
+  let [isLoading, setLoading] = useState(2);
+  let [randomStoriesData] = useState<string[]>([]);
+  let [randomStoriesContent] = useState<NewsFeedStory[]>([]);
```
Missing `set` functions for those two states.

jason-vasilev replied:
As they are arrays, I simply update them with `push`. I get "_____  is declared but its value is never read" warning if I would add the functions.

alordiel commented on src/NewsFeed/NewsFeed.tsx:
```diff
@@ -72,31 +74,29 @@ function NewsFeed() {
   }, []); // Empty dependency array to run effect only once on component mount
 
-  if (!hasLoaded) {
+  if (isLoading) {
```
If `isLoading` is 1 or 2 then this statement will be true and will execute its body, the statement will never reach to `isLoading` === 1.

alordiel commented on src/NewsFeed/NewsFeed.tsx:
```diff
@@ -36,16 +40,14 @@ function NewsFeed() {
-          // setRandomStories(randomStoryIds);
-          setRandomStoriesData(storyUrls);
+          setLoading(1); // let user know something is happening
           randomStoriesData = storyUrls;
```
Should be using `setRandomStoriesData`, if you don't get the data in the current rendering them problem might be [here](https://react.dev/reference/react/useState#updating-state-based-on-the-previous-state>). Same is valid for `randomStoriesContent`

alordiel commented on src/NewsFeed/NewsFeed.tsx:
```diff
@@ -38,79 +52,92 @@ class NewsFeed extends React.Component<
               (data) => {
-                this.randomStoriesContent.push(data);
+                randomStoriesContent.push(data);
+                return randomStoriesContent;
```
You don't need to return anything here as this value is not assigned anywhere. Same as below.

jason-vasilev committed:
isLoading rename, type and logic  update

jason-vasilev committed:
Migrate from CRA to Vite setup; yarn to npm; update dependancies

jason-vasilev committed:
Fix NewsFeed state mutation and add StrictMode-safe fetch

- Destructure setters from useState so updates actually trigger re-renders
- Replace direct array mutation (push/sort/reassign) with fresh setState values
- Add AbortController cleanup to cancel in-flight fetches on unmount, which also
  prevents duplicate items under React 18 StrictMode double-invocation
- Hoist pure helpers (getRandom, sort comparator) out of the component
- Drop the always-truthy array guard around the map

Made-with: Cursor

jason-vasilev committed:
Guard NewsCard karma fetch and hoist helpers

- Fetch author karma at most once per card via a useRef flag, instead of
  firing a new request on every mouseover/focus
- Abort the in-flight karma fetch on unmount to avoid setState on an
  unmounted component
- Hoist pure helpers (timeSince, imageId) out of the component and memoize
  timeAgo so it isn't recomputed on every unrelated re-render
- Type props as NewsCardProps instead of any, and add the optional url field
  to NewsFeedStory to reflect the HN API shape for non-poll stories

Made-with: Cursor

jason-vasilev committed:
Migrate SCSS from @import to @use/@forward

Dart Sass deprecated @import in favor of the module system, so replace
every @import in the stylesheets:

- main.scss now @forwards the generic partials so it still acts as a
  barrel that emits global CSS once and re-exports members downstream
- Partials that consume vars or mixins (_mixins, _typography, _grid)
  now explicitly @use 'vars' and 'mixins' with 'as *' to preserve
  unprefixed access, since @use no longer auto-globalizes members
- Component stylesheets switch from @import '../styles/main' to
  scoped @use of only the partials they actually need (vars, mixins,
  utilities). This also removes a pre-existing duplication where each
  component chunk was re-emitting all of main.scss's global CSS.

Verified via 'vite build' - no Sass deprecation warnings remain.

Made-with: Cursor

jason-vasilev committed:
Clean up pre-existing SCSS lint and CSS issues

- _normalize.scss: remove duplicated declarations left over from
  stripped vendor prefixes (text-size-adjust, height/width: fit-content),
  collapse 'monospace, monospace' to 'monospace', and replace the
  invalid '::input-placeholder' selector with '::placeholder'
- _normalize.scss: switch 'appearance: button' to 'appearance: auto'
  on button selectors and ::-webkit-file-upload-button - 'button' is
  a deprecated keyword, 'auto' is the modern equivalent
- _typography.scss: replace deprecated 'word-break: break-word' with
  'overflow-wrap: break-word' per the CSS Text spec
- _mixins.scss: rename stripUnit to strip-unit to satisfy
  scss/at-function-pattern (kebab-case convention)

Verified via 'npx vite build' (no Lightning CSS or Sass warnings)
and 'npx stylelint' (exits clean, 0 errors).

Made-with: Cursor

jason-vasilev committed:
Remove lock file
